core/codec/rle/rle_plus_decoding_stream.hpp

```diff
@@ -0,0 +1,180 @@
+     * @param output - decoded data
+     * @return Decoded stream
+     */
+    template <typename T>
+    RLEPlusDecodingStream &operator>>(std::set<T> &output) {
+      if ((content_.size() < Config::SMALL_BLOCK_LENGTH)
+          || (getSpan<uint8_t>(2) != 0)) {
+        throw errors::VersionMismatch();
+      }
+      magnitude_ = getSpan<uint8_t>(1) == 1;
+      T value = 0;
+      while (content_.find_next(index_ - 1)
+             != boost::dynamic_bitset<uint8_t>::npos) {
+        auto header = getSpan<uint8_t>(1);
+        if (header == 1) {
+          decodeSingleBlock(value, output);
+        } else if (header == 0) {
+          auto block_header = getSpan<uint8_t>(1);
+          if (block_header == 0) {
+            decodeLongBlock(value, output);
+          } else {
+            decodeSmallBlock(value, output);
+          }
+        }
+      }
+      constexpr size_t max_size = Config::OBJECT_MAX_SIZE / sizeof(T);
+      if (output.size() > max_size) {
+        throw errors::MaxSizeExceed();
+      }
+      return *this;
+    }
+
+   private:
+    size_t index_;   /**< Content's current index */
+    bool magnitude_; /**< Polarity of the current index */
+    boost::dynamic_bitset<uint8_t> content_; /**< Encoded data */
+
+    /**
+     * @brief Get bit span, represented as T
+     * @tparam T - type of value to return
+     * @param count - number of bits to use
+     * @return Bits [index_ : index_ + count], represented as T
+     */
+    template <typename T>
+    T getSpan(size_t count) {
+      T value{};
+      size_t shift = 0;
+      const size_t end = index_ + count;
+      if (content_.size() < end) {
+        throw errors::IndexOutOfBound();
+      }
+      for (size_t i = index_; i < end; ++i) {
+        T slice = content_.test(i) ? 1 : 0;
+        value |= slice << shift;
+        ++shift;
+      }
+      index_ += count;
+      return value;
+    }
+
+    /**
+     * @brief Unpack bytes to T
+     * @tparam T - type of the value to return
+     * @param data - packed bytes
+     * @return Unpacked data
+     */
+    template <typename T>
+    T unpack(const std::vector<uint8_t> &data) {
+      T value{};
+      size_t shift{};
+      const size_t max_shift = sizeof(T) * Config::BYTE_BITS_COUNT;
+      for (const auto &byte : data) {
+        if (shift > max_shift) {
+          throw errors::UnpackBytesOverflow{};
+        }
+        if (byte < Config::BYTE_SLICE_VALUE) {
+          value = value | (static_cast<T>(byte) << shift);
+          break;
+        }
+        value |= static_cast<T>(byte & Config::UNPACK_BYTE_MASK) << shift;
+        shift += Config::PACK_BYTE_SHIFT;
+      }
+      return value;
+    }
+
+    /**
+     * @brief Decode single RLE+ block
+     * @tparam T - type of the data to output
+     * @param current_value - index of the current number
+     * @param output - decoded data
+     */
+    template <typename T>
+    void decodeSingleBlock(T &current_value, std::set<T> &output) {
```

Comment:
Prefer using return values over output parameters.

Reply:
Google C++ Style Guide doesn't restrict using output parameters [#Output_Parameters](https://google.github.io/styleguide/cppguide.html#Output_Parameters)
If we prefer use return values here, we will get code like this:
```
// Method signature:
std::pair<T, std::set<T>> decodeSingleBlock(const T& current_value, const std::set<T> &output);

auto result = decodeSingleBlock(value, output);
value = result.first;
value = result.second;
```
Do we really need to choose return values here instead output parameters?

Reply:
We'd better had consistent code except it is absolutely necessary.

Reply:
Using return values here instead output parameters will lead to unnecessary copy operations: in first case we will create std::set<T> and next append it to the result, but in the second case we will directly emplace new values to the result.
Maybe, we have another way to implement this actions without extra copy operations and output parameters?

Reply:
Prefer return values instead of adding output parameters. 
There is no overhead in returning new set from this function. RVO will help to avoid redundant copy. All compilers do that

Reply:
Most likely I explained it unclear, please, look here again: it is not a RVO! 

In case of using return values:
1. We need to create new std::set and return it from function
2. Then we need to append this set to the result <- here is overhead copy operation

In case of using output parameters:
1. Emplace new values to the resulting set without additional copy operations